### PR TITLE
Add simple bag player examples

### LINKED ISSUE #1449: Add additional examples for reading ROSBag data and serialized messages
## Description

Currently, the ROSBag2 examples contain three different examples for writing data to a ROSBag via C++  or Python. However, there are no specific examples of reading data from ROSBags (neither for the standard reader or the sequential reader). In addition, the writing examples do not show how to serialize various messages before writing them to a bag file.
Inside the file under _rosbag2_tests/test/rosbag2_tests/test_rosbag2_cpp_api.cpp_, there are test cases for reading data and also for the message serialization. These helped me when using the ROSBag2 API, but in my opinion, a better idea is to create dedicated examples, because the unit tests should be meant solely for testing.

## Completion Criteria

- Example for reading ROSBag data using a simple reader (`rosbag2_cpp::Reader`)
- Example for reading ROSBag data using a sequential reader (`rosbag2_cpp::SequentialReader`)
- Example for writing ROSBag data using serialized messages (`rclcpp::SerializedMessage` and `rclcpp::Serialization<>`)

### PR DESCRIPTION
**Pull request by sangteak601**
Fixes #1449.

Added simple bag player examples in cpp and Python. I also tried to read or write bag files without specifying topic types. However, it became too complicated and was no longer a "simple" example, so I decided not to push the changes.

### CODE REVIEW

**Comment by sangteak601** [comment]
Hi @MichaelOrlov, I think I screwed up with reviewers. Do you mind fixing it and reviewing this PR ?

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: nit: against `simple_bag_recorder`.

```suggestion
  : Node("simple_bag_player")
```

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: Do we need to deserialize the message to be published? why can we not do `publish(rclcpp::SerializedMessage)`?

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: So this simple bag player does not care the timestamp in the bag file at all.

if user publishes `chatter` topic with 100Hz (10 msec) and record it with `simple_bag_recorder`, this will actually replay the messages with 1/10 frequency against the bag file timestamps.

i think this is no problem at all as example, but i think user needs to be aware of this situation. how about the adding comment somewhere here to explain this?

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py]:
```suggestion
# Copyright 2024 Open Source Robotics Foundation, Inc.
```

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (sangteak601) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: Thanks for the review.
Resolved.

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (sangteak601) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: Resolved

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py
Comment (sangteak601) [reply rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py]: Resolved

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (sangteak601) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: One of the requests from #1449 was to show how to serialize/deserialize messages. I also think it would be nice to include (de)serialization in the examples.

How about adding comments to inform that it is also possible to publish serialized messages?

**Comment by MichaelOrlov** [comment]
@sangteak601 @fujitatomoya I am curious how this PR overlaps with the existing tutorial and examples from https://docs.ros.org/en/rolling/Tutorials/Advanced/Reading-From-A-Bag-File-CPP.html ?

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (fujitatomoya) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: I see, as an example, that makes more sense. adding comment would be good to, i can leave this to you. thanks!

**Comment by fujitatomoya** [comment]
@MichaelOrlov 

> I am curious how this PR overlaps with the existing tutorial and examples from https://docs.ros.org/en/rolling/Tutorials/Advanced/Reading-From-A-Bag-File-CPP.html ?

that is what i found too.

i was thinking that https://docs.ros.org/en/rolling/Tutorials/Advanced/Reading-From-A-Bag-File-CPP.html can also be updated to point the source code from this example.

**Comment by MichaelOrlov** [comment]
> i was thinking that https://docs.ros.org/en/rolling/Tutorials/Advanced/Reading-From-A-Bag-File-CPP.html can also be updated to point the source code from this example.

I would prefer to make them aligned as much as possible but in the tutorial preferable to have the full text with source code instead of just a link on it.  

Also, I think that in the examples still unclear point that if the message always shall be deserialized or not before publishing.
To make it clear and show that it is optional I would suggest adding some class member variable or global variable with name like `message_needs_to_be_edit_before_send` and have an `if-else` statement showing two options for publishing. With deserialized and altering value before publishing and non-serialized publishing.

cc: @sangteak601

**Comment by sangteak601** [comment]
@MichaelOrlov Thanks for the review. I will make a modification.

Based on my knowledge(correct me if I'm wrong), we can also receive serialized messages directly. Therefore, perhaps we can modify the recorder example in a similar manner?

**Comment by sangteak601** [comment]
@fujitatomoya @MichaelOrlov I have addressed issues. Do you mind taking another look?

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: nothing to return?

```suggestion
      [this]() {this->timer_callback();}
```

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: i think it would be better for user to configure this flag via cli option to play around with this example package? unless, user needs to modify the source code and build it by them self from source, that requires more work for user?

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py:30]: if i am not mistaken, this is intentionally enabled as `sqlite3`? but this bag file can be also red from `simple_bag_player.cpp`, because it reads the meta data from the bag file and use the appropriate storage plugin if available? just checking this because user does not meet the failure during the tutorial.

**Review by fujitatomoya — COMMENTED** [review]
overall lgtm, but i have a few comments.

besides,

> I would prefer to make them aligned as much as possible but in the tutorial preferable to have the full text with source code instead of just a link on it.

i do agree with this. we need to update https://github.com/ros2/ros2_documentation accordingly to align with these samples.

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py
Comment (sangteak601) [reply rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py:30]: It was set as `sqlite3` in `simple_bag_recorder.py`. I don't know the reason exactly.
It is possible to read `sqlite3` files from `simple_bag_player.cpp`, but the opposite(reading `mcap` files from `simple_bag_player.py`) isn't.

Maybe we can delete `storage_id`?

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (sangteak601) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: Done

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (sangteak601) [reply rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp]: Add a ROS parameter `edit` to enable configuration.

**Comment by sangteak601** [comment]
> > I would prefer to make them aligned as much as possible but in the tutorial preferable to have the full text with source code instead of just a link on it.
> 
> i do agree with this. we need to update https://github.com/ros2/ros2_documentation accordingly to align with these samples.

I will update the documentation once this is merged.

File: rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py
Comment (fujitatomoya) [reply rosbag2_examples/rosbag2_examples_py/rosbag2_examples_py/simple_bag_player.py:30]: No, i think that is okay with current code. this also includes an example how to specify the storage id, i think that is fine.

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (fujitatomoya) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp:34]: comment for the documentation. this needs to be described how to enable with command line parameter argument. so that user can understand we can publish either publishing serialized data as it is or editing data before publishing in the example.

**Review by fujitatomoya — APPROVED** [review]

File: rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp
Comment (MichaelOrlov) [inline rosbag2_examples/rosbag2_examples_cpp/src/simple_bag_player.cpp:66]: @sangteak601 As far as I recall we need another type of publisher aka `GenericPublisher` to be able to publish serialized data.

**Comment by MichaelOrlov** [comment]
- Closing as stale PR in favor of the #1683

**Pull request closed without merging**
